pkg/util/routing/shared.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2022 The Codefresh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"fmt"
+
+	"github.com/codefresh-io/cli-v2/pkg/store"
+
+	"github.com/codefresh-io/cli-v2/pkg/util"
+	netv1 "k8s.io/api/networking/v1"
+	gatewayapi "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+type (
+	Controller interface {
+		Name() string
+		Decorate(route interface{})
+	}
+
+	CreateRouteOpts struct {
+		RuntimeName         string
+		Namespace           string
+		IngressClass        string
+		Hostname            string
+		InternalAnnotations map[string]string
+		ExternalAnnotations map[string]string
+		IngressController   Controller
+		GatewayName         string
+		GatewayNamespace    string
+	}
+)
+
+func CreateAppProxyRoute(opts *CreateRouteOpts, useGatewayAPI bool) (string, interface{}) {
+	var route interface{}
+	var routeName string
+
+	if useGatewayAPI {
+		httpRouteOpts := CreateHTTPRouteOptions{
+			Name:             opts.RuntimeName + store.Get().AppProxyIngressName,
+			Namespace:        opts.Namespace,
+			GatewayName:      opts.GatewayName,
+			GatewayNamespace: opts.GatewayNamespace,
+			Host:             opts.Hostname,
+			Rules: []HTTPRouteRule{
+				{
+					Path:        store.Get().AppProxyIngressPath,
+					PathType:    gatewayapi.PathMatchPathPrefix,
+					ServiceName: store.Get().AppProxyServiceName,
+					ServicePort: store.Get().AppProxyServicePort,
+				},
+			},
+		}
+
+		if opts.InternalAnnotations != nil {
+			httpRouteOpts.Annotations = make(map[string]string)
+			mergeAnnotations(httpRouteOpts.Annotations, opts.InternalAnnotations)
+		}
+
```

Comment:
Wouldn't it look better if you don't try to translate from `CreateRouteOpts` -> `CreateHTTPRouteOptions` -> `HTTPRoute` and instead just go straight from `CreateRouteOpts` -> `HTTPRoute`/`Ingress`? 
i.e. the `createHTTPRoute` and `CreateIngress` will take `CreateRouteOpts` and transform into their relative type.